### PR TITLE
SPEC 1515 Add spec test for MONGODB-AWS support

### PR DESCRIPTION
**Pull request by bazile-clyde**

### CODE REVIEW

**Comment by bazile-clyde** [comment]
I added a link to the testing guide in a comment on [DRIVERS-705](https://jira.mongodb.org/browse/DRIVERS-705)

File: source/auth/tests/mongodb-aws.rst [hunk 1]
```diff
@@ -0,0 +1,94 @@
...
+.. code-block:: bash
+
+  # with a session token
+  export AWS_ACCESS_KEY_ID=<AccessKeyId>
+  export AWS_SECRET_ACCESS_KEY=<SecretAccessKey>
+  export AWS_SESSION_TOKEN=<Token>
+
+  mongodb://localhost/?authMechanism=MONGODB-AWS
+|
+.. note:: no username, password or session token is passed into the URI. Drivers MUST check the environment variables listed above for these values. If the session token is set Drivers MUST use it.
```
Comment (jyemin) [inline source/auth/tests/mongodb-aws.rst]: Capitalize "no"

Comment (jyemin) [inline source/auth/tests/mongodb-aws.rst]: Nit: Looking at the file viewer, it seems like this doesn't format correctly

File: source/auth/tests/connection-string.json [hunk 2]
```diff
@@ -376,6 +376,63 @@
...
       "uri": "mongodb://:@localhost.com/",
       "valid": false
+    },
+    {
+      "description": "should recognise the mechanism (MONGODB-AWS)",
+      "uri": "mongodb://localhost/?authMechanism=MONGODB-AWS",
+      "hosts": null,
+      "valid": true,
+      "warning": false,
+      "auth": {
```
Comment (jyemin) [inline source/auth/tests/connection-string.json]: All the existing tests use a `credential` field that encompasses all this information.  Why is this different?

File: source/auth/tests/mongodb-aws.rst (hunk 1 above)
Comment (ShaneHarvey) [reply source/auth/tests/mongodb-aws.rst]: I think the problem might be the various `|` lines.

File: source/auth/tests/connection-string.json [hunk 3]
```diff
@@ -376,6 +376,63 @@
...
+        "password": "password",
+        "db": "$external"
+      },
+      "options": {
+        "authmechanism": "MONGODB-AWS"
+      }
+    },
+    {
+      "description": "should use username, password and session token if specified (MONGODB-AWS)",
+      "uri": "mongodb://user:password@localhost/?authMechanism=MONGODB-AWS&authMechanismProperties=AWS_SESSIONS_TOKEN:token",
```
Comment (ShaneHarvey) [inline source/auth/tests/connection-string.json]: Can we add some URL encoded characters to this TOKEN to make sure drivers decode in the right place?

Perhaps `AWS_SESSIONS_TOKEN:token%21%40%23%24%25%5E%26%2A%28%29_%2Betc` which decodes to `token!@#$%^&*()_+etc`. I had to fix a bug in Python which would have been easier to find with this kind of test.

Comment (ShaneHarvey) [reply source/auth/tests/connection-string.json]: Also the `authMechanismProperties` should be `AWS_SESSION_TOKEN` not `AWS_SESSIONS_TOKEN`.

File: source/auth/tests/connection-string.json [hunk 4]
```diff
@@ -376,6 +376,63 @@
       "description": "should throw an exception if no username/password provided (userinfo implies default mechanism)",
       "uri": "mongodb://:@localhost.com/",
       "valid": false
+    },
+    {
```
[the remark this answers is not in the document]
Comment (bazile-clyde) [reply source/auth/tests/connection-string.json]: Done

File: source/auth/tests/mongodb-aws.rst (hunk 1 above)
Comment (bazile-clyde) [reply source/auth/tests/mongodb-aws.rst]: Done

File: source/auth/tests/connection-string.json (hunk 4 above)
[the remark this answers is not in the document]
Comment (bazile-clyde) [reply source/auth/tests/connection-string.json]: Done

File: source/auth/tests/mongodb-aws.rst (hunk 1 above)
Comment (bazile-clyde) [reply source/auth/tests/mongodb-aws.rst]: Fixed. The issue was the bash syntax was wrong.

File: source/auth/tests/connection-string.json (hunk 2 above)
Comment (bazile-clyde) [reply source/auth/tests/connection-string.json]: Fixed

File: source/auth/tests/connection-string.json (hunk 3 above)
Comment (bazile-clyde) [reply source/auth/tests/connection-string.json]: Good idea. Added it for the username and password in a separate test as well.

**Pull request merged**